src/docs/services_and_contributions.md

```diff
@@ -4,42 +4,84 @@ title: Services and Contributions
 
-For instance, the `Navigator` widget needs access to a `FileSystem` to present
-folders and files in a tree. With DI the concretion of that `FileSystem`
-interface is not important to the `Navigator` widget. It can safely assume that
-an object consistent with the `FileSystem` interface is ready to use. In Theia,
-the used `FileSystem` concretion is just a proxy sending JSON-RPC messages to
-the backend, so it needs a particular configuration and treatment. The
-navigator doesn't need to care as it will get injected a fully working
-`FileSystem` instance.
+Theia uses the DI framework InversifyJS to wire up the different components including services and contribution points.
+DI decouples the consumers of services -- i.e. the dependencies of those consumers -- from the actual creation and retrieval of those services. As an example, if you want to use a service, you neither have to instantiate it, nor do you need to manually retrieve it from somewhere. Instead, the DI container injects the services on creation of your component. The DI container resolves the dependency for you and, if necessary, even instantiates it on the fly. With that, the consumer of services doesn’t need to worry where they come from and you can easily exchange the actual implementations of services later on without having to change the consumers. The DI container works based on some configuration you provide on startup through so-called container modules.
 
-Moreover, this decoupling of construction and use, allows extensions to provide
-their very specific implementations of e.g. a `FileSystem` if needed. Still
-without touching any users of the `FileSystem` interface.
+We will provide examples on how to use dependency injection in the sections “Services” and “Contributing to contribution points” below.
 
 DI is an integral part of Theia. Therefore, we highly recommend learning at
-least the basics of [Inversify.js](http://inversify.io/).
+least the basics of [InversifyJS](http://inversify.io/). For more details, please also refer to [this article on how dependency injection works in Theia](https://eclipsesource.com/blogs/2018/11/28/how-to-inversify-in-eclipse-theia/)
+
+## Using Services
+
+To use a service in Theia, you can get it injected as a dependency using DI. Dependencies are usually specified via the interface of the service you want to retrieve. This way, you even avoid a dependency on any specific implementation, the caller just knows the interface. This allows the component providing the implementation to seamlessly replace the implementation of a service. You can even override the existing default implementation of a service without breaking any service consumers.
 
-## Services
+Please note that InversifyJS technically uses Symbols to identify interfaces, because “Interface” is not a value in JavaScript. Therefore, service interfaces usually define a Symbol with exactly the same name as the interface itself.
```

Comment:
I think this paragraph is technically wrong and the whole "interface + symbol" thing had confused me for a long time. I think we should tell it like it really is: the DI container injects dependencies based on "service identifiers" that are put onto parameters and code via annotations (yes, class constructors can service as service identifiers, as well, but that's an implementation detail, really). Consumers of services should only rely on an interface when using a service so that service implementors are not locked into a specific implementation. By convention, the service identifier and the service interface have the same name.

Reply:
Hmm again. I agree, technically it is an identifier that you use. However, conceptionally it is the interface. In all DI frameworks in Java you can just use the interface itself as an identifier. Now do you tell someone: "You can get an instance of an interface identified by a String. You can use the interface itself and it will be translated to a string."
OR
"You can get an instance of an interface"
So again, do we focus on the technical implementation detail or the conceptional background? I prefer the second in these kind of articles.

Reply:
T.b.h, the whole: "symbol named like the interface" pattern does not make it easier unless you already understand InversifyJS: for example when you `bind(serviceIdentifier).toService(otherServiceIdentier)` it's just _way_ clearer IMO than when you're reading it as `bind(serviceIdentifier).toService(SomeInterface).

And no, conceptually it's not an interface. It's whatever is bound to that service identifier (basically a string). I can write ` bind(DebugService).to(QuickPickServiceImpl).inSingletonScope();` and the compiler will never complain. The _convention_ is to name the service identifier the same as service interface and to use service interfaces when consuming services. Saying that this is a convention will make more clear, not less.

I can only speak from personal experience, and the double identifiers for symbols and interfaces _have_ prevented me from really understanding what's happening with the DI used in Theia for a good while. I also repeatedly had to request changes in service binding when people couldn't bind to the interface, so they used the implementation class as the service identifier. All this is to convince you that while you may easily understand this pattern, not everyone seems to. I think there are a couple of points to be made here:

1. When writing a replaceable service, write a service interface and a class implementing that service
2. When using such a service (by getting it injected via DI), only rely on the interface, not the implementation
3. Define a symbol as the service identifier. Name that Symbol the same as the interface, following the pattern:

```
export const InterfaceName = Symbol('InterfaceName');
export interface InterfaceName {
...
```
4. Use the symbol as the service identifier when (re-)binding implementation to it

This is such an important pattern that I think it's worth fiddling with the language a bit to get it right.

Reply:
Adressed the commet @tsmaeder please check